ethereum/core/src/main/java/tech/pegasys/teku/core/operationvalidators/AttesterSlashingStateTransitionValidator.java

```diff
@@ -0,0 +1,111 @@
+ */
+
+package tech.pegasys.teku.core.operationvalidators;
+
+import static java.lang.Math.toIntExact;
+import static tech.pegasys.teku.core.operationvalidators.OperationInvalidReason.check;
+import static tech.pegasys.teku.core.operationvalidators.OperationInvalidReason.firstOf;
+import static tech.pegasys.teku.datastructures.util.AttestationUtil.is_slashable_attestation_data;
+import static tech.pegasys.teku.datastructures.util.AttestationUtil.is_valid_indexed_attestation;
+import static tech.pegasys.teku.datastructures.util.BeaconStateUtil.get_current_epoch;
+import static tech.pegasys.teku.datastructures.util.ValidatorsUtil.is_slashable_validator;
+
+import com.google.common.collect.Sets;
+import com.google.common.primitives.UnsignedLong;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.TreeSet;
+import tech.pegasys.teku.datastructures.operations.AttesterSlashing;
+import tech.pegasys.teku.datastructures.operations.IndexedAttestation;
+import tech.pegasys.teku.datastructures.state.BeaconState;
+
+public class AttesterSlashingStateTransitionValidator {
+
+  private List<UnsignedLong> indicesToSlash;
+
+  public Optional<OperationInvalidReason> validateSlashing(
+      final BeaconState state,
+      final AttesterSlashing attesterSlashing,
+      final Optional<Set<UnsignedLong>> intersectingIndices) {
+    indicesToSlash = new ArrayList<>();
+    IndexedAttestation attestation_1 = attesterSlashing.getAttestation_1();
+    IndexedAttestation attestation_2 = attesterSlashing.getAttestation_2();
+    return firstOf(
+        () ->
+            check(
+                is_slashable_attestation_data(attestation_1.getData(), attestation_2.getData()),
+                AttesterSlashingInvalidReason.ATTESTATIONS_NOT_SLASHABLE),
+        () ->
+            check(
+                is_valid_indexed_attestation(state, attestation_1).isSuccessful(),
+                AttesterSlashingInvalidReason.ATTESTATION_1_INVALID),
+        () ->
+            check(
+                is_valid_indexed_attestation(state, attestation_2).isSuccessful(),
+                AttesterSlashingInvalidReason.ATTESTATION_2_INVALID),
+        () -> {
+          boolean slashed_any = false;
+
+          Set<UnsignedLong> indices =
+              intersectingIndices.orElse(
+                  Sets.intersection(
+                      new TreeSet<>(
+                          attestation_1
+                              .getAttesting_indices()
+                              .asList()), // TreeSet as must be sorted
+                      new HashSet<>(attestation_2.getAttesting_indices().asList())));
+
+          for (UnsignedLong index : indices) {
+            if (is_slashable_validator(
+                state.getValidators().get(toIntExact(index.longValue())),
+                get_current_epoch(state))) {
+              indicesToSlash.add(index);
```

Comment:
Currently we use the same utilities for by using `OperationInvalidReason` in all operations. I agree it's a bit odd, but I don't know a simpler way to return this information from this calculation without breaking compatibility with the utilities we use such as `firstOf` and `check`.

Reply:
What about something like: 
```
public SlashingValidationResult validateSlashing(final BeaconState state, final AttesterSlashing slashing) {
   final List<UnsignedLong> slashedIndices = new ArrayList<>();
   Optional<OperationInvalidReason> failureReason = runChecks(state, slashing, slashedIndices);
   return failureReason.map(SlashingValidationResult::failedResult)
      .orElseGet(() -> SlashingValidationResult.successful(slashedIndices));
}

private Optional<OperationInvalidReason> runChecks(final BeaconState state, final AttesterSlashing attesterSlashing, final List<UnsignedLong> slashedIndices) {
  firstOf(...);
}
```

Reply:
I don't want to create a new `SlashingValidationResult` object because it might make deduplication of these state transition validators in the future harder, but I like the idea of passing in a List reference.